lib/logstash/filters/gramdict.rb

```diff
@@ -14,10 +15,12 @@
 # This class is useful for log file analysis, especially for identifying common patterns
 # and anomalies in log entries.
 class GramDict
-  def initialize
-    @tri_gram_dict = {}
-    @double_gram_dict = {}
-    @single_gram_dict = {}
+  def initialize(max_gram_dict_size)
+    # print "YASH YASH YASH YASH YASH ITS NEW GRAM DICT TIME YASH YASH YASH "
+
+    @tri_gram_dict = LruRedux::ThreadSafeCache.new(max_gram_dict_size)
```

Comment:
It seems that a single instance of gramdict is being shared across workers, despite what logstash's documentation seems to claim. To mitigate this, I'm just making the LruRedux use a `ThreadSafeCache` to fix race conditions across multiple workers. 

Clearly, there is a performance regression to doing this as we're now introducing a critical section. I'll add more details on how big that regression to this thread as I continue crunching the numbers. However, fundamentally, this allows the plugin to horizontally scale so the hit is worth it for the scalability metrics we really care about.

Reply:
Seems instead of taking 40 microseconds per log this way takes roughly 275 microseconds per log. I've also started another branch called `yash/fix_race_conditions_experiment` to see how deep I can get into fixing race conditions upon registering.